# Neurobiolog√≠a de la Conducta Social : RNA-seq
Dr. Jason Gallant, Michigan State University, USA

Dr. Vielka Salazar, Cape Breton University, Canada
****

## Getting Set Up:
All of what we need to do requires access to a remote server.  In order to complete the activity, you will need to bring your own laptop with the capability of connecting to via `ssh` to a server, and to run `x-windows` to look at graphical outputs.  Instructions for connecting to the server will be provided by your instructor.  You'll need to bring your laptop to participate in the course activities.  Please set up your laptop following *in advance*.

Click here for detailed setup [instructions](instructions.md)

## Schedule and Course Materials:
*April 1, 2019*
1. [Dr. Gallant's Lecture Notes](introduction.md) - Background on sequencing biochemistry and what RNA-seq is all about
2. [Dr. Salazar's Lecture Notes]() - Nanopore Sequencing

*April 2, 2019*
1. [Notes on Nanopore Sequencing](nanopore_notes.md)
2. [ONT Manual for Direct RNA Sequencing](images/direct-rna-sequencing-sqk-rna002-DRS_9080_v2_revB_22Nov2018.pdf)
3. [ONT Protocol for Direct RNA Sequencing](images/SQK-RNA002_protocol.pdf)

*April 3, 2019*
+ **14:30:** [Practical #1: Computing & Command Line Basics](computing.md) - A Quick Introduction to High Performance Computing and a Tutorial on the Command Line
+ **15:00:** [Practical #2](reads_and_qc.md) - An review of the RNA-seq workflow at "high speed", examining FASTQ Files and evaluating data quality
+ **15:30:**  Break - 15 min
+ **15:45:** [Practical #3](transcriptome_assembly.md) - An overview of de-novo transcriptome assembly using Trinity.  Examination of Trinity transcriptome and evaluation of outputs.
+ **16:15:** [Practical #4](alignment.md) - An overview of alignments, aligning Illumina reads and Oxford Nanopore reads
+ **16:45:** Break - 15 min
+ **17:00:** [Practical #5](nanopore_vs_illumina.md) - Comparing Illumina and Oxford Nanopore data.  Coverage, errors and other considerations.
+ **17:30:** Wrap-up and Review

### Sources and Resources:
+ [ANGUS Titus Brown - MSU](http://ged.msu.edu/angus/index.html)
+ [Eel Pond Tutorial](https://khmer-protocols.readthedocs.org/en/v0.8.4/mrnaseq/index.html)
+ [Trinity RNAseq](http://trinityrnaseq.sourceforge.net)
+ [Broad Institute Genome Bootcamp](http://www.broadinstitute.org/scientific-community/science/platforms/genome-sequencing/broadillumina-genome-analyzer-boot-camp)
+ [Colorado State Computational Biology and Genomics Workshop](https://dbsloan.github.io/TS2018/)
+ [Griffith Lab RNA-seq Tutorial Wiki](https://github.com/griffithlab/rnaseq_tutorial/wiki/Trinity-Assembly-And-Analysis)
